src/daft-functions/src/distance/cosine.rs

```diff
@@ -0,0 +1,140 @@
+use common_error::{DaftError, DaftResult};
+use daft_core::{
+    datatypes::{Field, Float64Array},
+    schema::Schema,
+    DataType, IntoSeries, Series,
+};
+use daft_dsl::{
+    functions::{ScalarFunction, ScalarUDF},
+    ExprRef,
+};
+use serde::{Deserialize, Serialize};
+use simsimd::SpatialSimilarity;
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct CosineDistanceFunction {}
+
+macro_rules! compute_cosine_distance {
+    ($source:expr, $query:expr, $type:ident) => {{
+        let query = &$query.fixed_size_list()?.flat_child;
+        let query = query.$type()?.as_slice();
+        $source
+            .into_iter()
+            .map(|list_opt| {
+                let list = list_opt?;
+                let list = list.$type().unwrap().as_slice();
+                let cosine = $type::cosine(&list, &query);
+                cosine
+            })
+            .collect::<Vec<_>>()
+    }};
+}
+
+#[typetag::serde]
+impl ScalarUDF for CosineDistanceFunction {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn name(&self) -> &'static str {
+        "cosine_distance"
+    }
+
+    fn evaluate(&self, inputs: &[Series]) -> DaftResult<Series> {
+        match inputs {
+            [source, query] => {
+                let source_name = source.name();
+                if query.len() != 1 {
+                    return Err(DaftError::ValueError(
+                        "Expected query to be a single value".to_string(),
+                    ));
+                }
+
+                let source = source.fixed_size_list()?;
+
+                let res = match query.data_type() {
+                    DataType::FixedSizeList(dtype, _) => match dtype.as_ref() {
+                        DataType::Int8 => {
+                            compute_cosine_distance!(source, query, i8)
```

Comment:
should we just use `simsimd` instead since it's already a dep?

Reply:
this is using `simsimd`. `$type::cosine(&list, &query);` the `cosine` method is added from `use simsimd::SpatialSimilarity;`.